File: app/src/main/java/com/stupidtree/hichat/ui/relationevent/RelationEventViewModel.java
package com.stupidtree.hichat.ui.relationevent;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Transformations;
import androidx.lifecycle.ViewModel;

import com.stupidtree.hichat.data.model.RelationEvent;
import com.stupidtree.hichat.data.model.UserLocal;
import com.stupidtree.hichat.data.repository.LocalUserRepository;
import com.stupidtree.hichat.data.repository.RelationRepository;
import com.stupidtree.hichat.ui.base.DataState;
import com.stupidtree.hichat.ui.base.Trigger;

import java.util.List;
import java.util.Objects;

public class RelationEventViewModel extends ViewModel {

    /**
     * 数据区
     */
    //数据本体：列表数据
    private LiveData<DataState<List<RelationEvent>>> listData;
    //Trigger：控制↑的刷新
    private MutableLiveData<Trigger> listDataController = new MutableLiveData<>();


    //状态数据：同意好友请求
    private LiveData<DataState<?>> responseResult;
    private MutableLiveData<ResponseFriendTrigger> responseFriendTrigger = new MutableLiveData<>();

    //状态数据：标记已读的结果
    private LiveData<DataState<Object>> markReadResult;
    //Trigger：控制↑的进行
    private MutableLiveData<Trigger> markReadController = new MutableLiveData<>();

    /**
     * 仓库区
     */
    LocalUserRepository localUserRepository;
    RelationRepository relationRepository;


    public RelationEventViewModel(){
        localUserRepository = LocalUserRepository.getInstance();
        relationRepository = RelationRepository.getInstance();
    }

    public LiveData<DataState<List<RelationEvent>>> getListData() {
        if(listData==null){
            listData = Transformations.switchMap(listDataController, input -> {
                if(!input.isActioning()){
                    return new MutableLiveData<>(new DataState<>(DataState.STATE.NOTHING));
                }
                UserLocal userLocal = localUserRepository.getLoggedInUser();
                if(userLocal.isValid()){
                    return relationRepository.queryMine(Objects.requireNonNull(userLocal.getToken()));
                }else{
                    return new MutableLiveData<>(new DataState<>(DataState.STATE.NOT_LOGGED_IN));
                }
            });
        }
        return listData;
    }

    public LiveData<DataState<?>> getResponseResult() {
        if(responseResult==null){
            responseResult = Transformations.switchMap(responseFriendTrigger, input -> {
                if(input.isActioning()){
                    UserLocal userLocal = localUserRepository.getLoggedInUser();
                    if(userLocal.isValid()){
                        return relationRepository.responseFriendRequest(Objects.requireNonNull(userLocal.getToken()),input.getEventId(),input.getAction());
                    }else{
                        return new MutableLiveData<>(new DataState<>(DataState.STATE.NOT_LOGGED_IN));
                    }
                }
                return new MutableLiveData<>(new DataState<>(DataState.STATE.NOTHING));
            });
        }
        return responseResult;
    }
    public LiveData<DataState<Object>> getMarkReadResult() {
        if(markReadResult==null){
            return Transformations.switchMap(markReadController, input -> {
                if(input.isActioning()){
                    UserLocal userLocal = localUserRepository.getLoggedInUser();
                    if(userLocal.isValid()){
                        return relationRepository.markRead(Objects.requireNonNull(userLocal.getToken()));
                    }else{
                        return new MutableLiveData<>(new DataState<>(DataState.STATE.NOT_LOGGED_IN));
                    }

                }
                return new MutableLiveData<>(new DataState<>(DataState.STATE.NOTHING));
            });
        }
        return markReadResult;
    }
    /**
     * 开始标记已读
     */
    public void startMarkRead(){
        markReadController.setValue(Trigger.getActioning());
    }
    public void startRefresh(){
        listDataController.setValue(Trigger.getActioning());
    }

    @Nullable
    public String getLocalUserId(){
        return localUserRepository.getLoggedInUser().getId();
    }

    public void responseFriendRequest(@NonNull String eventId, RelationEvent.ACTION action){
        responseFriendTrigger.setValue(ResponseFriendTrigger.getActioning(eventId,action));
    }
}

File: app/src/main/java/com/stupidtree/cloudliter/ui/relationevent/RelationEventViewModel.kt
package com.stupidtree.cloudliter.ui.relationevent

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Transformations
import androidx.lifecycle.ViewModel
import com.stupidtree.cloudliter.data.model.RelationEvent
import com.stupidtree.cloudliter.data.model.RelationEvent.ACTION
import com.stupidtree.cloudliter.data.repository.LocalUserRepository
import com.stupidtree.cloudliter.data.repository.RelationRepository
import com.stupidtree.cloudliter.data.repository.RelationRepository.Companion.instance
import com.stupidtree.cloudliter.ui.base.DataState
import com.stupidtree.cloudliter.ui.base.Trigger
import java.util.*

class RelationEventViewModel : ViewModel() {
    /**
     * 数据区
     */
    //数据本体：列表数据
    var listData: LiveData<DataState<List<RelationEvent>?>>? = null
        get() {
            if (field == null) {
                field = Transformations.switchMap(listDataController) { input: Trigger ->
                    if (!input.isActioning) {
                        return@switchMap MutableLiveData(DataState<List<RelationEvent>?>(DataState.STATE.NOTHING))
                    }
                    val userLocal = localUserRepository.loggedInUser
                    if (userLocal.isValid) {
                        return@switchMap relationRepository!!.queryMine(userLocal.token!!)
                    } else {
                        return@switchMap MutableLiveData(DataState<List<RelationEvent>?>(DataState.STATE.NOT_LOGGED_IN))
                    }
                }
            }
            return field
        }
        private set

    //Trigger：控制↑的刷新
    private val listDataController = MutableLiveData<Trigger>()

    //状态数据：同意好友请求
    var responseResult: LiveData<DataState<*>>? = null
        get() {
            if (field == null) {
                field = Transformations.switchMap(responseFriendTrigger) { input: ResponseFriendTrigger ->
                    if (input.isActioning) {
                        val userLocal = localUserRepository.loggedInUser
                        if (userLocal.isValid) {
                            return@switchMap relationRepository!!.responseFriendRequest(userLocal.token!!, input.eventId, input.action)
                        } else {
                            return@switchMap MutableLiveData<DataState<*>>(DataState<Any>(DataState.STATE.NOT_LOGGED_IN))
                        }
                    }
                    MutableLiveData<DataState<*>>(DataState<Any>(DataState.STATE.NOTHING))
                }
            }
            return field
        }
        private set
    private val responseFriendTrigger = MutableLiveData<ResponseFriendTrigger>()

    //状态数据：标记已读的结果
    val markReadResult: LiveData<DataState<Any?>>? = null
        get() = field
                ?: Transformations.switchMap(markReadController) { input: Trigger ->
                    if (input.isActioning) {
                        val userLocal = localUserRepository.loggedInUser
                        if (userLocal.isValid) {
                            return@switchMap relationRepository!!.markRead(userLocal.token!!)
                        } else {
                            return@switchMap MutableLiveData(DataState<Any?>(DataState.STATE.NOT_LOGGED_IN))
                        }
                    }
                    return@switchMap MutableLiveData(DataState<Any?>(DataState.STATE.NOTHING))
                }

    //Trigger：控制↑的进行
    private val markReadController = MutableLiveData<Trigger>()

    /**
     * 仓库区
     */
    var localUserRepository: LocalUserRepository
    var relationRepository: RelationRepository?

    /**
     * 开始标记已读
     */
    fun startMarkRead() {
        markReadController.value = Trigger.getActioning()
    }

    fun startRefresh() {
        listDataController.value = Trigger.getActioning()
    }

    val localUserId: String?
        get() = localUserRepository.loggedInUser.id

    fun responseFriendRequest(eventId: String, action: ACTION) {
        responseFriendTrigger.value = ResponseFriendTrigger.getActioning(eventId, action)
    }

    init {
        localUserRepository = LocalUserRepository.getInstance()
        relationRepository = instance
    }
}